discovery-consul/src/main/scala/akka/discovery/consul/ConsulSimpleServiceDiscovery.scala

```diff
@@ -0,0 +1,105 @@
+import com.orbitz.consul.model.ConsulResponse
+import ConsulSimpleServiceDiscovery._
+import akka.discovery.SimpleServiceDiscovery.{ Resolved, ResolvedTarget }
+import com.orbitz.consul.model.catalog.CatalogService
+import com.orbitz.consul.option.QueryOptions
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.FiniteDuration
+import scala.util.Try
+
+class ConsulSimpleServiceDiscovery(system: ActorSystem) extends SimpleServiceDiscovery {
+
+  private lazy val settings = ConsulSettings.get(system)
+  private lazy val consul =
+    Consul.builder().withHostAndPort(HostAndPort.fromParts(settings.consulHost, settings.consulPort)).build()
+
+  override def lookup(name: String, resolveTimeout: FiniteDuration): Future[SimpleServiceDiscovery.Resolved] = {
+    import system.dispatcher
+    Future.firstCompletedOf(
+      Seq(
+        after(resolveTimeout, using = system.scheduler)(
+          Future.failed(new TimeoutException("Future timed out!"))
+        ),
+        lookupInConsul(name)
+      )
+    )
+  }
+
+  private def lookupInConsul(name: String)(implicit executionContext: ExecutionContext): Future[Resolved] = {
+    val consulResult = for {
+      servicesWithTags <- getServicesWithTags
+      serviceIds = servicesWithTags.getResponse
+        .entrySet()
+        .asScala
+        .filter(e => e.getValue.contains(settings.applicationNameTagPrefix + name))
+        .map(_.getKey)
+      catalogServices <- Future.sequence(serviceIds.map(id => getService(id).map(_.getResponse.asScala.toList)))
+      resolvedTargets = catalogServices.flatten.toSeq.map(
+          catalogService => extractResolvedTargetFromCatalogService(catalogService))
+    } yield resolvedTargets
+    consulResult.map(targets => Resolved(name, scala.collection.immutable.Seq(targets: _*)))
+  }
+
+  private def extractResolvedTargetFromCatalogService(catalogService: CatalogService) = {
+    val port = catalogService.getServiceTags.asScala
+      .find(_.startsWith(settings.applicationAkkaManagementPortTagPrefix))
+      .map(_.replace(settings.applicationAkkaManagementPortTagPrefix, ""))
+      .flatMap { maybePort =>
+        Try(maybePort.toInt).toOption
+      }
+    ResolvedTarget(catalogService.getServiceAddress, Some(port.getOrElse(catalogService.getServicePort)))
+  }
+
+  private val getServicesWithTags = ((callback: ConsulResponseCallback[util.Map[String, util.List[String]]]) =>
+                                       consul.catalogClient().getServices(callback)).asFuture
+
+  private def getService(name: String) =
+    ((callback: ConsulResponseCallback[util.List[CatalogService]]) =>
+       consul.catalogClient().getService(name, QueryOptions.BLANK, callback)).asFuture
+
+}
+
+object ConsulSimpleServiceDiscovery {
+
+  implicit class ConsulResponseFutureDecorator[T](f: ConsulResponseCallback[T] => Unit) {
+    def asFuture: Future[ConsulResponse[T]] = {
+      val callback = new ConsulResponseFutureCallback[T]
+      f(callback)
```

Comment:
could `f` throw? Let's make sure that then the Future is failed rather than the asFuture throwing

Reply:
I think I did what you asked for, but please take another look